core-c-bridge/src/worker.rs

```diff
@@ -282,36 +359,53 @@ impl<SK: SlotKind + Send + Sync> temporal_sdk_core_api::worker::SlotSupplier
     type SlotKind = SK;
 
     async fn reserve_slot(&self, ctx: &dyn SlotReservationContext) -> SlotSupplierPermit {
-        let (tx, rx) = oneshot::channel();
         let ctx = Self::convert_reserve_ctx(ctx);
-        let tx = Box::into_raw(Box::new(tx)) as *mut libc::c_void;
+        let completion_ctx = Arc::new(SlotReserveCompletionCtx {
+            state: AtomicCell::new(SlotReserveOperationState::Pending),
+            notify: Notify::new(),
+        });
         unsafe {
-            let _drop_guard = CancelReserveGuard {
-                token_src: ctx.token_src,
-                callback: (*self.inner.0).cancel_reserve,
-            };
-            ((*self.inner.0).reserve)(&ctx, tx);
-            rx.await.expect("reserve channel is not closed")
+            let inner = &*self.inner.0;
+            (inner.reserve)(&ctx, Arc::into_raw(completion_ctx.clone()), inner.user_data);
+        }
+        let mut guard = CancelReserveGuard {
+            slot_supplier: self,
+            completion_ctx,
+            completed: false,
+        };
+        // if the future is dropped before this await resolves, the guard is dropped which triggers cancellation
+        guard.completion_ctx.notify.notified().await;
+        guard.completed = true;
+        match guard.completion_ctx.state.load() {
+            SlotReserveOperationState::Completed(permit_id) => {
+                SlotSupplierPermit::with_user_data::<usize>(permit_id.get())
+            }
+            other => panic!("Unexpected slot reservation state: expected Completed, got {other:?}"),
         }
     }
 
     fn try_reserve_slot(&self, ctx: &dyn SlotReservationContext) -> Option<SlotSupplierPermit> {
         let ctx = Self::convert_reserve_ctx(ctx);
-        let permit_id = unsafe { ((*self.inner.0).try_reserve)(&ctx) };
+        let permit_id = unsafe { ((*self.inner.0).try_reserve)(&ctx, (*self.inner.0).user_data) };
         if permit_id == 0 {
             None
         } else {
-            Some(SlotSupplierPermit::with_user_data(permit_id))
+            Some(SlotSupplierPermit::with_user_data::<usize>(permit_id))
         }
     }
 
     fn mark_slot_used(&self, ctx: &dyn SlotMarkUsedContext<SlotKind = Self::SlotKind>) {
         let ctx = SlotMarkUsedCtx {
             slot_info: Self::convert_slot_info(ctx.info().downcast()),
-            slot_permit: ctx.permit().user_data::<usize>().copied().unwrap_or(0),
+            slot_permit: ctx
+                .permit()
+                .user_data::<usize>()
+                .copied()
+                .expect("Failed to read custom slot supplier permit ID"),
```

Comment:
Small stylistic note, but I think it's better when `expect` calls read as what should be expected. In this case, something like "Custom slot supplier permit ID should always be a usize"